### PR TITLE
Improve error messages for invalid identities

### PR DESCRIPTION
**Pull request by kimorris27**
### Which issue this PR addresses:

https://redhat.atlassian.net/browse/ARO-27289

### What this PR does / why we need it:

If we get a 401, 403, 404, or 429 getting a MI, give the error to the user - otherwise return an internal server error. This is a follow-up on https://github.com/Azure/ARO-RP/pull/4161. I think there were some comments there that didn't get addressed before merge.

Example below of what the error looks like from customer perspective at the CLI for a 403. Do we want to add more custom messaging to clarify that the issue is the MSI lacking permission or leave it like this?

```
$ az aro create <options>
...
(InvalidPlatformWorkloadIdentity) GET https://management.azure.com/subscriptions/REDACTED/resourceGroups/kimorris-v4-westeurope/providers/Microsoft.ManagedIdentity/userAssignedIdentities/image-registry
--------------------------------------------------------------------------------
RESPONSE 403: 403 Forbidden
ERROR CODE: AuthorizationFailed
--------------------------------------------------------------------------------
{
  "error": {
    "code": "AuthorizationFailed",
    "message": "The client 'REDACTED' with object id 'REDACTED' does not have authorization to perform action 'Microsoft.ManagedIdentity/userAssignedIdentities/read' over scope '/subscriptions/REDACTED/resourceGroups/kimorris-v4-westeurope/providers/Microsoft.ManagedIdentity/userAssignedIdentities/image-registry' or the scope is invalid. If access was recently granted, please refresh your credentials."
  }
}
--------------------------------------------------------------------------------

Target: .properties.platformWorkloadIdentityProfile.platformWorkloadIdentities["image-registry"]
```

### Test plan for issue:

Unit tests and E2E

### Is there any documentation that needs to be updated for this PR?

No

### How do you know this will function as expected in production? 

E2E

### CODE REVIEW

File: pkg/util/azureerrors/error.go [hunk 1]
```diff
@@ -172,6 +172,34 @@ func IsStatusConflictError(err error) bool {
...
+		return detailedErr.StatusCode == http.StatusForbidden
+	}
+
+	var responseError *azcore.ResponseError
+	if errors.As(err, &responseError) {
+		return responseError.StatusCode == http.StatusForbidden
+	}
+
+	return false
+}
```
[the remark this answers is not in the document]
Comment (kimorris27) [reply pkg/util/azureerrors/error.go:175]: Rejecting this. I like the current pattern where the name of each helper function reflects in a very human-readable way which type of error we're checking for, and it's not like we're going to end up with infinitely many of these helpers.

File: pkg/cluster/platformworkloadidentities.go [hunk 2]
```diff
@@ -44,11 +44,10 @@ func (m *manager) platformWorkloadIdentityIDs(ctx context.Context) error {
 
 		identityDetails, err := m.userAssignedIdentities.Get(ctx, resourceId.ResourceGroupName, resourceId.Name, &armmsi.UserAssignedIdentitiesClientGetOptions{})
 		if err != nil {
-			if azureerrors.Is4xxError(err) {
-				m.log.Error(err)
-				return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidPlatformWorkloadIdentity, fmt.Sprintf(`.properties.platformWorkloadIdentityProfile.platformWorkloadIdentities["%s"]`, operatorName), fmt.Sprintf("platform workload identity '%s' is invalid", operatorName))
+			if azureerrors.IsStatusUnauthorizedError(err) || azureerrors.IsStatusForbiddenError(err) || azureerrors.IsNotFoundError(err) {
+				return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidPlatformWorkloadIdentity, fmt.Sprintf(`.properties.platformWorkloadIdentityProfile.platformWorkloadIdentities["%s"]`, operatorName), err.Error())
```
Comment (iormungand) [inline pkg/cluster/platformworkloadidentities.go:48]: The previous code logged errors server-side before returning a CloudError. Now only a CloudError is returned, which will only result in an INFO level error being logged, pointing to the CloudError.

This will make debugging these customer issues more difficult. The previous error at least would have pointed to platform workload identity, and could be found during a smaller set of log lines focused on ERROR level. Now seeing the error will require looking through the much larger set of INFO and above lines, and the message will not mention platform workload identity, but just reference an Azure error, which as far as I can tell will be a much more opaque URL.

There are multiple other consequences of downgrading these errors from ERROR to INFO. Strongly recommend at least adding back the `m.log.Error`.

Unfortunately, the other (non 401, 403, 404) errors get better logging, because returning as a fmt.Errorf means they will bubble up to the step runner and have the extra context of which step failed. I don't see a way to do that for the errors in question if we have to return a CloudError, without somewhat rearchitecting the error handling.

File: pkg/cluster/platformworkloadidentities.go [hunk 3]
```diff
@@ -44,11 +44,10 @@ func (m *manager) platformWorkloadIdentityIDs(ctx context.Context) error {
 
 		identityDetails, err := m.userAssignedIdentities.Get(ctx, resourceId.ResourceGroupName, resourceId.Name, &armmsi.UserAssignedIdentitiesClientGetOptions{})
 		if err != nil {
-			if azureerrors.Is4xxError(err) {
-				m.log.Error(err)
-				return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidPlatformWorkloadIdentity, fmt.Sprintf(`.properties.platformWorkloadIdentityProfile.platformWorkloadIdentities["%s"]`, operatorName), fmt.Sprintf("platform workload identity '%s' is invalid", operatorName))
+			if azureerrors.IsStatusUnauthorizedError(err) || azureerrors.IsStatusForbiddenError(err) || azureerrors.IsNotFoundError(err) {
```
Comment (iormungand) [inline pkg/cluster/platformworkloadidentities.go]: Narrowing the scope here of 4xx errors to only 401, 403, 404 runs the risk of other errors not surfacing because they now fall through to internal server error. One example of this would be a 429, rate limit error. Should we give them some sort of CloudError handling as well - even if not as detailed of a treatment as these?

File: pkg/util/azureerrors/error.go [hunk 4]
```diff
@@ -172,6 +172,34 @@ func IsStatusConflictError(err error) bool {
 	return false
 }
 
+func IsStatusUnauthorizedError(err error) bool {
```
Comment (iormungand) [inline pkg/util/azureerrors/error.go:169]: Might want to change the IsNotFoundError to a IsStatusNotFoundError, to make it consistent with the others, and as the not found is inferred from status. If that would cause too many code changes elsewhere, then maybe the new ones could be renamed to IsUnauthorizedError and IsForbiddenError.

File: pkg/util/azureerrors/error.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (iormungand) [reply pkg/util/azureerrors/error.go:175]: If I may be a little cheeky here, unrejecting this.

At least consider this. I also like the current pattern with the name of the helper function is very readable. But, what if we inverted the suggestion? That is, have the named helper functions wrap a function that covers the duplication. Because it is a good point that with the current way, it is easier to introduce an inconsistency.

In particular, I think this might be helpful because there is a potential pre-existing bug in the error handling function above. Is4xxError only checks azcore.ResponseError, but not autorest.DetailedError. This means the previous code wouldn't have caught 4xx errors from autorest based calls.

Maybe this isn't happening, but if the named helper functions wrapped functions called something like isStatusCodeErrorForAny(statusCode StatusCode), and isStatusCodeErrorForResponse(statusCode StatusCode), the code would at least make it clear that for a specific error type, it was deliberate that only the response was being checked. I think though in general, without a specific reason against, that we should probably check both in all cases, even if it seems like autorest is being gradually deprecated.

Additionally, there is an even more complex error checking flow below, in ResourceGroupNotFound, that all of the other errors might benefit by using.

File: pkg/cluster/platformworkloadidentities_test.go [hunk 5]
```diff
@@ -37,6 +39,10 @@ func TestPlatformWorkloadIdentityIDs(t *testing.T) {
 	identityBarResourceId := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ManagedIdentity/userAssignedIdentities/%s", subscriptionId, clusterRG, identityBarName)
 	identityBarClientId := "0ba40ba4-0ba4-0ba4-0ba4-0ba40ba40ba4"
 	identityBarObjectId := "1ba41ba4-1ba4-1ba4-1ba4-1ba41ba41ba4"
+	fooTarget := `.properties.platformWorkloadIdentityProfile.platformWorkloadIdentities["foo"]`
+	unauthorizedErr := azruntime.NewResponseError(&http.Response{StatusCode: http.StatusUnauthorized})
```
Comment (iormungand) [inline pkg/cluster/platformworkloadidentities_test.go:43]: Might be useful to add in a RawResponse containing a body with a more realistic error message, so we can assert that this message surfaces where we want it to.

File: pkg/cluster/platformworkloadidentities.go (hunk 2 above)
Comment (iormungand) [inline pkg/cluster/platformworkloadidentities.go:48]: Should authn/authz errors be changed to a BadRequest status? Would that make permissions errors harder to trace?

File: pkg/cluster/platformworkloadidentities_test.go [hunk 6]
```diff
@@ -121,7 +127,76 @@ func TestPlatformWorkloadIdentityIDs(t *testing.T) {
...
+							},
+						},
+					},
+				},
+			},
+			userAssignedIdentitiesClientMocks: func(mock *mock_armmsi.MockUserAssignedIdentitiesClient) {
+				mock.EXPECT().Get(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes().
+					Return(armmsi.UserAssignedIdentitiesClientGetResponse{}, notFoundErr)
+			},
+			wantErr: api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidPlatformWorkloadIdentity, fooTarget, notFoundErr.Error()).Error(),
```
Comment (iormungand) [inline pkg/cluster/platformworkloadidentities_test.go:200]: There should be at least one test verifying fallthrough behavior for the newly separated 4xx errors that are not 401/403/404. Some other edge cases could possibly be tested as well.

**Review by iormungand — CHANGES_REQUESTED** [review]
This PR will deliver better messages to customers, but there are some logging issues that need to be handled, and some testing issues that could be improved.

My answer to the question posed in the PR summary is, for a 403 error in particular, this could be user actionable - if they haven't registered the RP, if there are policies blocking RP access to resources, if the identity is in a different subscription without the right configuration, and probably others I'm not aware of. Anything we can surface to help the customer realize the issue and fix it themselves is one less ticket in the queue.

File: pkg/cluster/platformworkloadidentities.go (hunk 3 above)
Comment (cadenmarchese) [reply pkg/cluster/platformworkloadidentities.go]: We are handling 429s in https://github.com/Azure/ARO-RP/pull/4785 already, however, we wouldn't want to return a 429 to the customer (probably) since that is an internal Azure platform issue that is not actionable to them. In this PR, the goal is to focus on the actionable "invalid identities" errors, so we can defer to #4785 for the 429s.

File: pkg/cluster/platformworkloadidentities.go (hunk 2 above)
Comment (cadenmarchese) [reply pkg/cluster/platformworkloadidentities.go:48]: I don't think so - we're handling an instance where the request to the API is fine and correct, but the permissions (which are not part of the ARM payload) are incorrect.

Comment (cadenmarchese) [reply pkg/cluster/platformworkloadidentities.go:48]: The cloud errors are already logged in RP in addition to being returned to customer, so m.log.Error(err) that was removed was a redundant/duplicate log message.

File: pkg/util/azureerrors/error.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kimorris27) [reply pkg/util/azureerrors/error.go:175]: Going point by point:

> At least consider this. I also like the current pattern with the name of the helper function is very readable. But, what if we inverted the suggestion? That is, have the named helper functions wrap a function that covers the duplication. Because it is a good point that with the current way, it is easier to introduce an inconsistency.
>
>In particular, I think this might be helpful because there is a potential pre-existing bug in the error handling function above. Is4xxError only checks azcore.ResponseError, but not autorest.DetailedError. This means the previous code wouldn't have caught 4xx errors from autorest based calls.

That's fair. I'm down to make a shared "internal helper" that covers the duplication.

>  I think though in general, without a specific reason against, that we should probably check both in all cases, even if it seems like autorest is being gradually deprecated.

Yep, I agree that we should continue to check both error types in all cases for now.

> Additionally, there is an even more complex error checking flow below, in ResourceGroupNotFound, that all of the other errors might benefit by using.

I think I'd stop short of doing this for this PR. I don't think it necessarily provides more value than giving the full Azure error to the customer, and enumerating all of the cases we'd want to check for and how we'd want to customize the error messages is a good chunk of extra work.

File: pkg/cluster/platformworkloadidentities_test.go (hunk 6 above)
Comment (cadenmarchese) [reply pkg/cluster/platformworkloadidentities_test.go:200]: I agree, but only if it's handled in a separate test that is testing the separate non-40* handling path. Since this PR focuses on 40* errors the test seems fine to me.

**Review by cadenmarchese — APPROVED** [review]

File: pkg/cluster/platformworkloadidentities.go (hunk 2 above)
Comment (kimorris27) [reply pkg/cluster/platformworkloadidentities.go:48]: Yep, and I confirmed the step runner logs these errors at error level for us: https://github.com/Azure/ARO-RP/blob/f8b36ba6c46dc80751b13a41efdb68685093ad4d/pkg/util/steps/runner.go#L98

File: pkg/cluster/platformworkloadidentities_test.go (hunk 5 above)
Comment (kimorris27) [reply pkg/cluster/platformworkloadidentities_test.go:43]: I don't think this is necessary for now since we're just checking error codes. If we do end up making changes later on to check the messages in addition to the codes, we could revisit this.

File: pkg/cluster/platformworkloadidentities.go (hunk 2 above)
Comment (kimorris27) [reply pkg/cluster/platformworkloadidentities.go:48]: Even though it's an auth error at the level of the managed identity API, it's still a bad request from the ARO-RP's perspective. We expect some conditions to be met so that we can fulfill the API request (the cluster identity's permissions over the platform identities), and we reject the request as invalid if the customer hasn't ensured those conditions.

The 403, forbidden, authorization failed, etc. will also be included when the RP logs the error, making it easy to search by those things.

File: pkg/cluster/platformworkloadidentities_test.go (hunk 6 above)
Comment (kimorris27) [reply pkg/cluster/platformworkloadidentities_test.go:200]: There's already a case that handles generic errors: `error - unknown error when fetching details from ARM`.

I can't think of what other edge cases you'd be referring to, but I'd consider adding them if you identify the ones you have in mind.

File: pkg/cluster/platformworkloadidentities.go (hunk 3 above)
Comment (kimorris27) [reply pkg/cluster/platformworkloadidentities.go]: I can go ahead and add in a call to `IsRetryableError`. Since the platform identities are customer-owned resources, it's good for them to know that the issue is 429. Otherwise we're obfuscating an issue on the customer's end by returning an internal server error for that case.

File: pkg/util/azureerrors/error.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kimorris27) [reply pkg/util/azureerrors/error.go:175]: Factored out the type checking stuff. I refrained from touching some of the other more complex error handling functions in the file to avoid scope creep.

**Review by slawande2 — COMMENTED** [review]
nit: `Is4xxError` doesn't have any callers anymore. Doesn't need to be in this PR, but we should clean it up in a follow-up to avoid leaving dead code.

**Review by slawande2 — APPROVED** [review]

**Comment by kimorris27** [comment]
> nit: `Is4xxError` doesn't have any callers anymore. Doesn't need to be in this PR, but we should clean it up in a follow-up to avoid leaving dead code.

I meant to do this while I was pushing other changes earlier and then forgot! Good catch. I went ahead and took care of it.

**Review by iormungand — APPROVED** [review]
I thoroughly approve of the refactors that help improve code readability.

I understand and agree with the other responses regarding error flow. Error handling, in general, seems to be one of the areas especially lacking documentation.

This is a solid change at this point.
